### PR TITLE
Add Gradle's 'assemble' check to catch the distribution generation issues

### PR DESCRIPTION
**Pull request by reta**
### Description
Add Gradle's 'assemble' check to catch the distribution generation issues

### Related Issues
Related to https://github.com/opensearch-project/OpenSearch/issues/8903

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@andrross @dblock @nknize would love to hear your thoughts on the matter. It turned out we don't really run any checks on distributions (and sadly, had a regression sneaked in https://github.com/opensearch-project/OpenSearch/issues/8903). 

I would like to suggest to add one more `assemble` check to  make sure the distributions are properly built across the set of available operating systems (Linux / MacOS / Windows). That would have helped us to catch the regression early on (the checks are failing exactly because of this problem). 

Thank you folks.

**Comment by owaiskazi19** [comment]
@reta what do you think of integrating this check in the same workflow as https://github.com/opensearch-project/OpenSearch/pull/8486? This workflow checks for the compatibility by running `./gradlew assemble` in plugins.

**Comment by reta** [comment]
> @reta what do you think of integrating this check in the same workflow as #8486? This workflow checks for the compatibility by running `./gradlew assemble` in plugins? Sequence I think should be like:
> 
>     1. run assemble core to check for distribution generation issues
> 
>     2. run compatibility check with the plugins
> 
> 
> in the same workflow file.

Thanks @owaiskazi19, the workflows are somewhat different, I would not necessarily bundle them together, but that is not the main reason. The `assemble` command is run on 3 different OSes whereas compatibility check just need one single pass.

**Comment by andrross** [comment]
> @andrross @dblock @nknize would love to hear your thoughts on the matter. It turned out we don't really run any checks on distributions (and sadly, had a regression sneaked in #8903).
> 
> I would like to suggest to add one more `assemble` check to make sure the distributions are properly built across the set of available operating systems (Linux / MacOS / Windows). That would have helped us to catch the regression early on (the checks are failing exactly because of this problem).
> 
> Thank you folks.

@reta I like the intent here for sure! In the name of frugality, are we duplicating the compilation that is being done for these platforms in `precommit` or `check`? If so, is that worth worrying about? In theory it seems like we could piggyback `assemble` on the `check` action for ubuntu and the `precommit` action for macos and windows.

**Comment by reta** [comment]
> @reta I like the intent here for sure! In the name of frugality, are we duplicating the compilation that is being done for these platforms in `precommit` or `check`?

Thanks a lot @andrross , yes, we do duplicate compilation and javadoc

>  In theory it seems like we could piggyback assemble on the check action for ubuntu and the precommit action for macos and windows.

The problem with `check` is that it runs off the Jenkins, whereas macos & windows precommits are Github actions, the later could be combined with `assemble`. What we could try is:
 - bring back `precommit` for Linux
 - but exclude it from Gradle check (Jenkins), still compilation would be repeated sadly 

What do you think?

File: .github/workflows/precommit.yml [hunk 1]
```diff
@@ -1,19 +1,29 @@
...
+name: Gradle Precommit and Asssemble
 on: [pull_request]
 
 jobs:
   precommit:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [windows-latest, macos-latest]  # precommit on ubuntu-latest is run as a part of the gradle-check workflow
+        os: [ubuntu-latest, windows-latest, macos-latest]  # precommit on ubuntu-latest is run as a part of the gradle-check workflow
```
Comment (andrross) [inline .github/workflows/precommit.yml:9]: Can probably remove the "# precommit on ubuntu-latest is run as a part of the gradle-check workflow" comment now. It was meant to explain why ubuntu wasn't included in this list.

**Review by andrross — APPROVED** [review]

File: .github/workflows/precommit.yml (hunk 1 above)
Comment (reta) [reply .github/workflows/precommit.yml]: Certainly, thanks @andrross !

**Pull request merged**